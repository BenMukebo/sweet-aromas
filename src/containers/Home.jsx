import React, { useState, useEffect } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import HashLoader from 'react-spinners/HashLoader';
import _ from 'lodash';

import axios from 'axios';
import Filter from '../components/Filter';
import Header from '../components/Header';
import Search from '../components/Search';
import Sort from '../components/Sort';
import RecipesList from './RecipesList';

import {
  changeFilter,
  loadCategoriesAsync,
  loadRecipesAsync,
} from '../store/actions';
import paginate from '../components/utils/paginate';

const Home = (props) => {
  const [selectedCategory, setSelectedCategory] = useState('All');
  const [query, setQuery] = useState('');
  const [orderColumn, setOrderColumn] = useState('None');
  const pageSize = 6;
  const [currentPage, setCurrentPage] = useState(1);
  const { loadCategories, loadRecipes } = props;

  const [currentUser, setCurrentUser] = useState('');
  const baseUrl = 'https://sweetaromas.herokuapp.com';

  const handleRequest = async () => {
    const config = {
      headers: {
        Authorization: localStorage.getItem('token'),
      },
    };

    try {
      const response = await axios.get(`${baseUrl}/users`, config);
      const { data } = response;
      console.log('Show', data.users);
      setCurrentUser(data);
      // localStorage.setItem('token', authorization);

      // const { token } = res.data;
    } catch (error) {
      console.log(error);
    }
  };

  useEffect(() => {
    loadCategories();
    loadRecipes();
    handleRequest();
  }, [loadCategories, loadRecipes]);
  // componentDidMount() {

  // }

  const handleSearchRecipe = (e) => {
    const { changeFilter } = props;
    changeFilter('All');
    setQuery({ query: e.target.value, selectedCategory: 'All' });
  };

  const handleSeletectCategory = (category) => {
    const { changeFilter } = props;
    setSelectedCategory({ selectedCategory: category, query: '', currentPage: 1 });
    changeFilter(category);
  };

  const handleChangeOrderColumn = (orderColumn) => {
    setOrderColumn({ orderColumn });
  };

  const handlePageChange = (currentPage) => {
    setCurrentPage({ currentPage });
  };

  const renderFilteredRecipes = (recipes) => {
    const { filter } = props;
    if (filter === 'All') return recipes;
    return recipes.filter((recipe) => recipe.strCategory.includes(filter));
  };

  const renderSearchedRecipes = (recipes) => {
    if (query === '') return recipes;

    return recipes.filter(
      (recipe) => recipe.strMeal.toLowerCase().includes(query.toLocaleLowerCase())
        || recipe.strArea.toLowerCase().includes(query.toLocaleLowerCase()),
    );
  };

  const renderSortedRecipes = (recipes) => {
    if (orderColumn === 'None') return recipes;
    const iteratee = orderColumn === 'Name' ? 'strMeal' : 'strArea';
    return _.orderBy(recipes, iteratee, 'ASC');
  };

  const {
    categories,
    recipes,
    loadingCategories,
    loadingRecipes,
    loadCategoriesError,
    loadRecipesError,
  } = props;
  const filteredRecipes = renderFilteredRecipes(recipes);
  const searchedRecipes = renderSearchedRecipes(filteredRecipes);
  const sortedRecipes = renderSortedRecipes(searchedRecipes);
  const pagedRecipes = paginate(sortedRecipes, currentPage, pageSize);
  return (
    <div>
      <Header />
      <div className="home-content-main-area d-flex">
        <div className="side-bar-wrapper">
          {loadingCategories && (
            <div className="loading-spinner-wrapper d-flex flex-center">
              <HashLoader color="#e0aea6" loading={loadingRecipes} size={70} />
            </div>
          )}
          {!loadingCategories && (
            <Filter
              categories={categories}
              selectedCategory={selectedCategory}
              onChangeFilter={handleSeletectCategory}
              error={loadCategoriesError}
            />
          )}
        </div>
        <div className="main-home-content">
          <div className="header d-flex flex-center flex-between">
            <Search query={query} onChange={handleSearchRecipe} />
            <Sort
              onChangeSortColumn={handleChangeOrderColumn}
              activeColumn={orderColumn}
            />
          </div>
          {loadingRecipes && (
            <div className="loading-spinner-wrapper d-flex flex-center">
              <HashLoader color="#e0aea6" loading={loadingRecipes} size={70} />
            </div>
          )}
          {!loadingRecipes && (
            <RecipesList
              recipes={pagedRecipes}
              itemsCount={sortedRecipes && sortedRecipes.length}
              currentPage={currentPage}
              pageSize={pageSize}
              onPageChange={handlePageChange}
              error={loadRecipesError}
            />
          )}
        </div>
      </div>
    </div>
  );
};

Home.defaultProps = {
  loadRecipesError: '',
  loadCategoriesError: '',
};

Home.propTypes = {
  loadCategories: PropTypes.func.isRequired,
  categories: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  loadRecipes: PropTypes.func.isRequired,
  recipes: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  changeFilter: PropTypes.func.isRequired,
  filter: PropTypes.string.isRequired,
  loadingCategories: PropTypes.bool.isRequired,
  loadingRecipes: PropTypes.bool.isRequired,
  loadCategoriesError: PropTypes.string,
  loadRecipesError: PropTypes.string,
};

const mapStateToProps = (state) => ({
  recipes: state.recipes.list,
  filter: state.filter,
  categories: state.categories.list,
  loadingCategories: state.categories.loading,
  loadingRecipes: state.recipes.loading,
  loadCategoriesError: state.categories.error,
  loadRecipesError: state.recipes.error,
});

const mapDispatchToProps = {
  loadCategories: () => loadCategoriesAsync(),
  loadRecipes: () => loadRecipesAsync(),
  changeFilter: (filter) => changeFilter(filter),
};

export default connect(mapStateToProps, mapDispatchToProps)(Home);
