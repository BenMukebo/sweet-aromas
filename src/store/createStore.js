import { createStore } from 'redux';

export default () => createStore(reducer);
