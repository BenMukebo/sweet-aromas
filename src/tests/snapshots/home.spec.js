import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { render, screen } from '@testing-library/react';
import createStore from '../../store/createStore';

import Home from '../../containers/Home';

const store = createStore();

describe('Home', () => {
  it('should render correctly', () => {
    const tree = renderer
      .create(
        <BrowserRouter>
          <Provider store={store}>
            <Home />
          </Provider>
        </BrowserRouter>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should renders recipes link', () => {
    render(
      <BrowserRouter>
        <Provider store={store}>
          <Home />
        </Provider>
      </BrowserRouter>,
    );
    const linkElement = screen.getByText(/recipes/i);
    expect(linkElement).toBeInTheDocument();
  });

  it('should not render an unknown link', () => {
    render(
      <BrowserRouter>
        <Provider store={store}>
          <Home />
        </Provider>
      </BrowserRouter>,
    );
    expect(() => screen.getByText(/happy/i)).toThrow();
  });

  it('should throw an error if looking for an unexisting element', () => {
    render(
      <BrowserRouter>
        <Provider store={store}>
          <Home />
        </Provider>
      </BrowserRouter>,
    );
    expect(() => screen.getByText(/food/i)).toThrow();
  });

  it('should not render an unknown link', () => {
    render(
      <BrowserRouter>
        <Provider store={store}>
          <Home />
        </Provider>
      </BrowserRouter>,
    );
    expect(() => screen.getByText(/happy/i)).toThrow();
  });
});
