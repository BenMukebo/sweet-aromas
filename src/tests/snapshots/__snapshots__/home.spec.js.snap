// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should render correctly 1`] = `
<div>
  <header
    className="app-header"
  >
    <div
      className="header-main-container d-flex flex-center"
    >
      <div
        className="header-wrapper flex-unit d-flex"
      >
        <div
          className="hero-area d-flex flex-center flex-column"
        >
          <h1
            className="hero-text"
          >
            Recipes
            <br />
            Catalogue
          </h1>
          <p
            className="hero-paragraph"
          >
            Lorem ipsum dolor sit amet, consectetur adipisicing elit. Voluptate quae nesciunt porro maiores quod aliquam vero dicta animi ipsa assumenda.
          </p>
          <a
            className="how-they-work d-flex flex-center"
            href="/"
            onClick={[Function]}
          >
            How they work
            <svg
              className="hero-arrow-icon"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clipRule="evenodd"
                d="M10.146 4.646a.5.5 0 01.708 0l3 3a.5.5 0 010 .708l-3 3a.5.5 0 01-.708-.708L12.793 8l-2.647-2.646a.5.5 0 010-.708z"
                fillRule="evenodd"
              />
              <path
                clipRule="evenodd"
                d="M2 8a.5.5 0 01.5-.5H13a.5.5 0 010 1H2.5A.5.5 0 012 8z"
                fillRule="evenodd"
              />
            </svg>
          </a>
        </div>
        <div
          className="hero-image-wrapper"
        >
          <img
            alt="hiro"
            className="hiro-image"
            src="hero-illustration.png"
          />
        </div>
      </div>
    </div>
  </header>
  <div
    className="home-content-main-area d-flex"
  >
    <div
      className="side-bar-wrapper"
    >
      <div
        className="loading-spinner-wrapper d-flex flex-center"
      >
        <span
          className="css-1u252oj"
        >
          <span
            className="css-rqmy7v"
          />
          <span
            className="css-17y5dvi"
          />
        </span>
      </div>
    </div>
    <div
      className="main-home-content"
    >
      <div
        className="header d-flex flex-center flex-between"
      >
        <div
          className="search-bar-main-container d-flex flex-center"
        >
          <svg
            className="search-icon"
            fill="none"
            height="1em"
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="11"
              cy="11"
              r="8"
            />
            <line
              x1="21"
              x2="16.65"
              y1="21"
              y2="16.65"
            />
          </svg>
          <input
            className="search-input"
            onChange={[Function]}
            placeholder="Search recipes by name or area..."
            type="text"
            value=""
          />
        </div>
        <div
          className="sort-main-wrapper d-flex flex-center flex-between"
        >
          <span>
            Sort By: 
          </span>
          <a
            className="selected-filter-method d-flex flex-center"
            href="#f"
            onClick={[Function]}
          >
            None
            <svg
              className="sort-arrow"
              fill="none"
              height="1em"
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polyline
                points="6 9 12 15 18 9"
              />
            </svg>
          </a>
          <ul
            className="sort-method-list"
          >
            <li>
              <a
                data-sort="None"
                href="#f"
                onClick={[Function]}
                value="None"
              >
                None
              </a>
            </li>
            <li>
              <a
                data-sort="Name"
                href="#f"
                onClick={[Function]}
                value="Name"
              >
                Name
              </a>
            </li>
            <li>
              <a
                data-sort="Area"
                href="#f"
                onClick={[Function]}
                value="Area"
              >
                Area
              </a>
            </li>
          </ul>
        </div>
      </div>
      <div
        className="loading-spinner-wrapper d-flex flex-center"
      >
        <span
          className="css-1u252oj"
        >
          <span
            className="css-rqmy7v"
          />
          <span
            className="css-17y5dvi"
          />
        </span>
      </div>
    </div>
  </div>
</div>
`;
